### PR TITLE
Add find command to list ACLs containing the given ID in their formulae

### PR DESCRIPTION
**Pull request by rattus-rattus-rattus**

### CODE REVIEW

File: plugins/db_manager.py [hunk 1]
```diff
@@ -174,6 +175,62 @@ async def acl_show(ctx: Context, acl: str) -> None:
...
+                        ", ".join(format("{!i}", acl) for acl in matched_acls))
+    if len(indirect_acls):
+        output += format("ACLs indirectly referencing {} {!i}:\n- {}", type, id,
+                        "\n- ".join(format("{!i} via <@&{}>", acl, role) for acl, role in indirect_acls))
+    if output == "":
+        output = format("No ACLs referencing {} {!i}", type, id)
+
+    await ctx.send(output, allowed_mentions=AllowedMentions.none())
+
+def find_in_acl(data: ACLData | None, type: str, id: int) -> bool:
```
Comment (mniip) [inline plugins/db_manager.py:220]: This should be an abstract method on `ACL`, rather than walking the data representation.

Comment (mniip) [reply plugins/db_manager.py:220]: As pyright says `X | Y` is union syntax only supported in python 3.10+, we target 3.9

File: plugins/db_manager.py [hunk 2]
```diff
@@ -174,6 +175,62 @@ async def acl_show(ctx: Context, acl: str) -> None:
 
     await ctx.send(format("{!b:yaml}", yaml.dump(data)))
 
+@acl_command.command("find")
+@privileged
+async def acl_find(ctx: Context, type: str, id: int) -> None:
```
Comment (mniip) [inline plugins/db_manager.py:180]: You could have `type: Literal["role", "channel", "category", "user"]` and the library would check that (and provide a better help signature for the command).

File: plugins/db_manager.py [hunk 3]
```diff
@@ -174,6 +175,62 @@ async def acl_show(ctx: Context, acl: str) -> None:
...
+        raise UserError("Can only be used in a guild")
+
+    if type not in ["role", "channel", "category", "user"]:
+        raise UserError("Invalid type, must be one of `user`, `role`, `channel`, or `category`")
+    if type == "role" and ctx.guild.get_role(id) is None:
+        raise UserError(format("No role with ID: {!i}", id))
+    if type in ("channel", "category") and ctx.guild.get_channel(id) is None:
+        raise UserError(format("No channel with ID: {!i}", id))
+    if type == "user" and ctx.guild.get_member(id) is None:
+        raise UserError(format("No user with ID: {!i}", id))
```
Comment (mniip) [inline plugins/db_manager.py:193]: Don't see why these checks have to be carried out at all. I appreciate the ergonomics of specifying a user and have the bot automatically look up their roles, but I don't think it should prevent you from doing a lookup on e.g. a deleted role.

File: plugins/db_manager.py [hunk 4]
```diff
@@ -174,6 +175,62 @@ async def acl_show(ctx: Context, acl: str) -> None:
...
+    else:
+        user = ctx.guild.get_member(id)
+
+    all_acls: Set[str] = set(name for ty, name in bot.acl.conf if ty == "acl")
```
Comment (mniip) [inline plugins/db_manager.py:197]: Consider a set comprehension: `{name for ty, name in bot.acl.conf if ty == "acl"}`

File: plugins/db_manager.py [hunk 5]
```diff
@@ -174,6 +175,62 @@ async def acl_show(ctx: Context, acl: str) -> None:
...
+                            for acl in all_acls 
+                            for role in user.roles 
+                            if find_in_acl(bot.acl.conf["acl", acl], "role", role.id))
+
+    if len(matched_acls):
+        output += format("ACLs directly referencing {} {!i}: {}\n", type, id, 
+                        ", ".join(format("{!i}", acl) for acl in matched_acls))
+    if len(indirect_acls):
+        output += format("ACLs indirectly referencing {} {!i}:\n- {}", type, id,
+                        "\n- ".join(format("{!i} via <@&{}>", acl, role) for acl, role in indirect_acls))
```
Comment (mniip) [inline plugins/db_manager.py:214]: instead of `"\n- " + "\n- ".join(... for ...)` why not `"".join("\n- " + ... for ...)`

Comment (mniip) [reply plugins/db_manager.py:214]: To mention a role use `{!M}`

**Comment by jkynaston** [comment]
refactored code to use abstract method on `ACL`; updated readme to match signature; added some additional string manipulation to "prettify" output

File: plugins/db_manager.py [hunk 6]
```diff
@@ -174,6 +175,50 @@ async def acl_show(ctx: Context, acl: str) -> None:
...
+                                and ACL.parse(datum).refers_to(type, id)}
+    indirect_acls: Dict[str,List[int]] = {}
+    user = ctx.guild.get_member(id)
+    output = ""
+
+    if user is not None:
+        for acl in all_acls:
+            for role in user.roles:
+                if (datum := bot.acl.conf[("acl", acl)]) is not None and ACL.parse(datum).refers_to("role", role.id):
+                    indirect_acls.setdefault(acl, []).append(role.id)
```
Comment (jkynaston) [inline plugins/db_manager.py:197]: may be a cleaner/more pythonic way to do this

File: plugins/db_manager.py [hunk 7]
```diff
@@ -174,6 +175,50 @@ async def acl_show(ctx: Context, acl: str) -> None:
...
+        format_type = " {!c}"
+    elif type == "category":
+        format_type = " {!i}"
+
+    if len(matched_acls):
+        output += format("ACLs directly referencing {}" + format_type + ": {}\n", 
+                         type, id, ", ".join(format("{!i}", acl) for acl in matched_acls))
+
+    if len(indirect_acls):
+        output += format("ACLs indirectly referencing {!m}:", id)
```
Comment (mniip) [inline plugins/db_manager.py:213]: Your definition of "A indirectly referencing X" seems to be "X is a member and X has role Y and A references Y", which... is it really that useful a notion? What about channels and categories?

File: bot/acl.py [hunk 8]
```diff
@@ -107,6 +107,10 @@ def evaluate(self, user: Optional[Union[Member, User]], channel: Optional[Messag
         ) -> EvalResult:
         raise NotImplemented
 
+    @abstractmethod
+    def refers_to(self, type: str, id: int) -> bool:
```
Comment (mniip) [inline bot/acl.py:111]: The control flow in the command could be greatly streamlined I think, if it were `references(self) -> Set[int]` instead. I don't think a "type" parameter is necessary, but if it is, it should be at least a `Literal[...]`

**Pull request closed without merging**